core/metacling/src/TCling.cxx

```diff
@@ -5831,6 +5831,108 @@ Int_t TCling::AutoLoad(const std::type_info& typeinfo, Bool_t knowDictNotLoaded
+            }
+            status = 1;
+         }
+         else {
+            Error("TCling::AutoLoad",
+                  "failure loading library %s for %s", lib, cls);
+         }
+      }
+      delete tokens;
+   }
+
+   return status;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Iterate through the data member of the class (either through the TProtoClass
+// or through Cling) and trigger, recursively, the loading the necessary libraries.
+Int_t TCling::DeepAutoLoadImpl(const char *cls)
+{
+   Int_t status = ShallowAutoLoadImpl(cls);
+   if (status) {
+
+      // This routine should be called only from AutoLoad which has already
+      // taken the main ROOT lock so this should not have any race condition.
+      // If the lock is removed from AutoLoad, a spin lock should be introduced here.
+      // Note that it is actually alright if another thread is populating this
+      // set since we can then exclude both the infinite recursion (the main goal)
+      // and duplicate work.
+      static std::set<std::string> gClassOnStack;  
+      auto insertResult = gClassOnStack.insert(std::string(cls));
+      if (insertResult.second) {
+         // Now look through the TProtoClass to load the required library/dictionary
+         TProtoClass *proto = TClassTable::GetProto(cls);
+         if (proto) {
+            for(auto element : proto->GetData()) {
+               const char *subtypename = element->GetTypeName();
+               if (!element->IsBasic() && !TClassTable::GetDictNorm(subtypename)) {
+                  // Failure to load a dictionary is not (quite) a failure load
+                  // the top-level library.  If we return false here, then
+                  // we would end up in a situation where the library and thus
+                  // the dictionary is loaded for "cls" but the TClass is
+                  // not created and/or marked as unavailable (in case where
+                  // AutoLoad is called from TClass::GetClass).
+                  (void) DeepAutoLoadImpl(subtypename);
+               }
+            }
+         } else {
+            auto classinfo = gInterpreter->ClassInfo_Factory(cls);
+            if (classinfo && gInterpreter->ClassInfo_IsValid(classinfo)
+                && !(gInterpreter->ClassInfo_Property(classinfo) & kIsEnum))
+            {
+               DataMemberInfo_t *memberinfo = gInterpreter->DataMemberInfo_Factory(classinfo);
+               while (gInterpreter->DataMemberInfo_Next(memberinfo)) {
+                  auto membertypename = TClassEdit::GetLong64_Name(gInterpreter->TypeName(gInterpreter->DataMemberInfo_TypeTrueName(memberinfo)));
+                  if (!(gInterpreter->DataMemberInfo_TypeProperty(memberinfo) & ::kIsFundamental)
+                      && !TClassTable::GetDictNorm(membertypename.c_str()))
+                  {
+                     (void)DeepAutoLoadImpl(membertypename.c_str());
```

Comment:
Shouldn't the return result of this call affect the final return result?

Reply:
No, see the comment in the previous call to DeepAutoLoadImpl (line 5902)